synapse_spamcheck_badlist/BadListFilter.py

```diff
@@ -0,0 +1,212 @@
+        except Exception as e:
+            logger.warn("We CANNOT check links! %s" % e)
+            self._can_we_check_links = False
+        return self._can_we_check_links
+        
+    async def can_we_check_md5(self) -> bool:
+        """
+            Check whether the MD5 database exists, caching the result.
+        """
+        if self._can_we_check_md5 is not None:
+            return self._can_we_check_md5
+        if self._md5_table is None:
+            self._can_we_check_md5 = False
+            return False
+        try:
+            def interaction(db):
+                db.execute("SELECT md5 FROM %s LIMIT 1" % self._md5_table)
+            await self._api.run_db_interaction("Check whether we can check md5", interaction)
+            self._can_we_check_md5 = True
+            logger.info("We can check md5!")
+        except:
+            logger.warn("We CANNOT check md5!")
+            self._can_we_check_md5 = False
+        return self._can_we_check_md5
+
+    async def check_event_for_spam(self, event) -> bool:
+        if event["type"] != "m.room.message":
+            # We only filter messages.
+            return False
+
+        # If it's a text content, look for links.
+        if event["content"]["msgtype"] == "m.text":
+            if not await self.can_we_check_links():
+                return False
+
+            # Check for links in text, both unformatted and formatted.
+            #
+            # We always lower-case the text, as the IWF database is lowercase.
+            for text in [event.get("content", {}).get("body", ""), event.get("content", {}).get("formatted_body", "")]:
+                for match in self._linkifier.match(text.lower()) or []:
+                    link = re.sub(self._scheme_re, "", match.url)
+                    is_bad_link = await self._api.run_db_interaction("Check link against evil db", _db_is_bad_link, self._links_table, link)
+                    if is_bad_link:
+                        logger.info("Rejected bad link")
+                        return True
+
+        # If it's a file, download content, extract hash.
+        elif event["content"]["msgtype"] == "m.file":
+            if not await self.can_we_check_md5():
+                return False
+
+            match = self._mxc_re.match(event["content"]["url"])
+            server_name = None
+            media_id = None
+            if match != None:
+                server_name = match.group('server_name')
+                media_id = match.group('media_id')
+                response = None
+                try:
+                    url = "%s/_matrix/media/r0/download/%s/%s" % (
+                            self._base_url,
+                            urlquote(server_name),
+                            urlquote(media_id)
+                    )
+                    response = await self._api.http_client.request("GET", url)
+                    # FIXME: What do we do if we were rate-limited?
+                except Exception as e:
```

Comment:
This error checking does not seem to work, I think you want to `continue` after logging.

Reply:
I think we want to bailout and `return False`, no?

Reply:
I don't know what behavior is best, it depends if you want to fail closed or opened (and potentially if you want retries). You might also want to check other links.

Reply:
Other links? There can be more than one file in a `m.file` message?

Reply:
Oops, sorry I got this confused with checking the links in a standard message!